src/av_stubs.c

```diff
@@ -730,76 +728,67 @@ CAMLprim value ocaml_av_select_stream(value _stream)
 }
 
 
-static value decode_packet(av_t * av, stream_t * stream, AVPacket * packet, AVFrame * frame)
+static inline int decode_packet(av_t * av, stream_t * stream, AVPacket * packet, AVFrame * frame)
 {
   AVCodecContext * dec = stream->codec_context;
-  value frame_kind = 0;
-  int ret = AVERROR_INVALIDDATA;
+  int ret = 0;
+
+  caml_release_runtime_system();
 
   if(dec->codec_type == AVMEDIA_TYPE_AUDIO ||
      dec->codec_type == AVMEDIA_TYPE_VIDEO) {
   
-    ret = avcodec_send_packet(dec, packet);
-  
-    if (ret >= 0 || ret == AVERROR(EAGAIN) || ret == AVERROR_EOF) {
-  
-      if (ret >= 0) packet->size = 0;
-  
-      // decode frame
-      ret = avcodec_receive_frame(dec, frame);
-    }
-  
-    if(dec->codec_type == AVMEDIA_TYPE_AUDIO) {
-      frame_kind = PVV_Audio;
-    }
-    else {
-      frame_kind = PVV_Video;
+    if (!av->frames_pending) {
```

Comment:
```suggestion
    // Assumption: each time this function is called with `frames_pending == 0`, 
    // a fresh packet is also provided and no packet otherwise.
    if (!av->frames_pending) {
```